### PR TITLE
core: Allow implicit module addition

### PR DESCRIPTION
**Pull request by compor**
This PR allows the implicit addition of a `builtin.module` operation if it is missing.
The functionality is:

- No `builtin.module` is added if there is already a single top-level `builtin.module`
- A `builtin.module` is added when:
  - A single top-level operation has been parsed and it is not `builtin.module`
  - Multiple operations have been parsed

In other words,

```mlir
%0 = "test.op"() : () -> ()
```

is being returned as

```mlir
"builtin.module"() ({
  %0 = "test.op"() : () -> ()
}) : () -> ()
```

In more detail:

- Addition of `--no-implicit-module` command-line option similar to `mlir-opt`:
  - by default set to `False`
  - when set to `True`, it has the behaviour that `xdsl-opt` exhibited so far (i.e., a single top-level `builtin.module` to be present
- Adds a boolean flag to control this behaviour in `MLContext`
- Adds a few tests to improve coverage and exercise the new and old functionality

### CODE REVIEW

File: xdsl/ir/core.py [hunk 1]
```diff
@@ -59,6 +59,7 @@ class MLContext:
     """Contains structures for operations/attributes registration."""
 
     allow_unregistered: bool = field(default=False)
+    allow_implicit_module: bool = field(default=True)
```
[the remark this answers is not in the document]
Comment (webmiche) [reply xdsl/ir/core.py:62]: From a conceptual point of view, it feels like this should not be part of the MLContext. This seems mostly like something specific to parsing, the actual module does not really care.

I understand though that this is just a very easy way to do this as you pass around the ctx anyways. Couldn't we also solve this with just a single argument to the parse_module function? Is this really that much uglier?

**Review by webmiche — COMMENTED** [review]
Thanks a lot, good to get this.

File: xdsl/ir/core.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (compor) [reply xdsl/ir/core.py:62]: I did that to start with, but it was rather ugly and the propagation was deep. As I said, open to suggestions.

[the remark this answers is not in the document]
Comment (webmiche) [reply xdsl/ir/core.py:62]: Essentially, line 313 of xdsl_opt_main should be the only change, right?

[the remark this answers is not in the document]
Comment (compor) [reply xdsl/ir/core.py:62]: Let me have another look, maybe I missed this.

[the remark this answers is not in the document]
Comment (compor) [reply xdsl/ir/core.py:62]: Thanks for the suggestion, it's much better; I think I somehow missed this although it was the first thing I tried.

**Review by webmiche — APPROVED** [review]
Nice! I prefer this design as well :+1:

**Review by superlopuh — APPROVED** [review]

File: xdsl/xdsl_opt_main.py [hunk 2]
```diff
@@ -267,6 +267,14 @@ def register_all_arguments(self, arg_parser: argparse.ArgumentParser):
             action="store_true",
             help="Allow the parsing of unregistered dialects.",
         )
+
+        arg_parser.add_argument(
```
Comment (math-fehr) [inline xdsl/xdsl_opt_main.py:271]: Same here, I would remove the argument, as we always want this option to be True.

**Review by math-fehr — APPROVED** [review]
Nice! I think I would remove the option so `builtin.module` are always added.
MLIR also does not have the option, so I would just copy that!

**Comment by compor** [comment]
> Nice! I think I would remove the option so `builtin.module` are always added. MLIR also does not have the option, so I would just copy that!

@math-fehr are we sure? I'm looking at LLVM 16 and it does? it also skips the module when printing MLIR.

2023-07-12-084012_908x29_scrot

File: xdsl/parser/core.py [hunk 3]
```diff
@@ -107,15 +107,34 @@ def __init__(
         self.forward_block_references = dict()
         self.forward_ssa_references = dict()
 
-    def parse_module(self) -> ModuleOp:
-        op = self.parse_optional_operation()
+    def parse_module(self, allow_implicit_module: bool = True) -> ModuleOp:
```
[the remark this answers is not in the document]
Comment (compor) [reply xdsl/parser/core.py:110]: No specific use case, it was mostly looking at `mlir-opt` which propagates the option down.

File: xdsl/parser/core.py [hunk 4]
```diff
@@ -107,15 +107,34 @@ def __init__(
...
+
+            if not isinstance(parsed_op, ModuleOp):
+                self._resume_from(0)
+                self.raise_error("builtin.module operation expected", 0)
+
+            module_op = parsed_op
+        else:
+            parsed_ops: list[Operation] = []
+
+            while (parsed_op := self.parse_optional_operation()) is not None:
```
[the remark this answers is not in the document]
Comment (compor) [reply xdsl/parser/core.py]: Done and added this as a test case.

**Review by math-fehr — APPROVED** [review]
My bad! I think I missed the option when I was looking at it yesterday, so that's good to me now!

**Pull request merged**